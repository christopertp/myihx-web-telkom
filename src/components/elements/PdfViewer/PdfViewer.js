import React, { useEffect, useRef } from 'react';
import { getDocument, GlobalWorkerOptions } from 'pdfjs-dist';
import pdfjsWorker from 'pdfjs-dist/build/pdf.worker.entry';
import PropTypes from 'prop-types';
import { remToPx } from '../../../utils/unit';
import styles from './styles.scoped.css';

export default function PdfViewer(props) {
  const { className, url } = props;
  const classes = [styles.root, className, 'loading'].filter(Boolean).join(' ');
  const ref = useRef(null);

  useEffect(() => {
    (async () => {
      if (!url) {
        return;
      }

      GlobalWorkerOptions.workerSrc = pdfjsWorker;

      const pdf = await getDocument(url).promise;
      const pages = [...Array.from({ length: pdf.numPages }).keys()];

      const clientWidth = ref.current.clientWidth - (2 * remToPx(1));
      ref.current.innerHTML = '';

      pages.reduce(async (acc, i) => {
        await acc;

        const page = await pdf.getPage(i + 1);
        const { width } = page.getViewport({ scale: 1 });
        const viewport = page.getViewport({ scale: clientWidth / width });

        const canvas = document.createElement('canvas');
        const devicePixelRatio = window.devicePixelRatio || 1;
        const transform = [ devicePixelRatio, 0 , 0, devicePixelRatio, 0, 0];

        canvas.width = viewport.width * devicePixelRatio;
        canvas.height = viewport.height * devicePixelRatio;
        ref.current.appendChild(canvas);

        const canvasContext = canvas.getContext('2d');
        const renderTask = page.render({ canvasContext, transform, viewport });

        return renderTask.promise;
      }, Promise.resolve());
      ref.current.classList.remove('loading');
    })();
  }, [url]);

  return <div className={classes} ref={ref} />;
}

PdfViewer.defaultProps = {
  className: '',
  url: '',
};

PdfViewer.propTypes = {
  className: PropTypes.string,
  url: PropTypes.string,
};
