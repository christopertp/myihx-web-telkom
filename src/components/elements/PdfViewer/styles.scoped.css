.root {
  background-color: #d8d8d8;
  max-height: 37.5rem;
  overflow-y: auto;
  padding: 1rem;
}

.root:empty {
  height: 20rem;
}

.root > canvas {
  margin-bottom: 1rem;
  width: 100%;
}

.root > canvas:last-child {
  margin-bottom: 0;
}

@media only screen and (max-width: 48em) {
  .root {
    width: 100%;
  }
}
