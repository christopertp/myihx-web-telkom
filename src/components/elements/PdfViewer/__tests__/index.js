import React, { useRef } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import PdfViewer from '../PdfViewer';

document.createElement = () => ({
  getContext: jest.fn(),
});

describe('src/components/elements/PdfViewer', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<PdfViewer />);
    expect(tree).toMatchSnapshot();
  });

  test('render with url', async () => {
    const current = {
      appendChild: jest.fn(),
      classList: {
        remove: jest.fn(),
      },
    };
    useRef.mockImplementationOnce(() => ({ current }));

    await PdfViewer({ url: 'tes' });

    useRef.mockImplementationOnce(() => ({ current }));

    delete global.devicePixelRatio;
    global.devicePixelRatio = 0;

    await PdfViewer({ url: 'tes' });

    expect(current.classList.remove).toHaveBeenCalledWith('loading');
  });
});
