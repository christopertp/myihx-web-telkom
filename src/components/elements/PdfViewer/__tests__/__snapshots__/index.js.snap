// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/PdfViewer render 1`] = `
<div
  className="loading"
/>
`;
